File: problems/problem_004/typescript/solution.test.ts
import { zigzagConversion } from './solution';
import { test, describe, expect } from 'bun:test';

describe('zigagConversion', () => {
    type zigzagConverionTest = [string, number, string];

    const tests: zigzagConverionTest[] = [
        ['HELLOPROGRAMMING', 4, 'HRMEPOMILOGANLRG'],
        ['PROBLEMCHALLENGE', 5, 'PHRCAEOMLGBELNLE'],
        ['ABCDE', 2, 'ACEBD'],
    ];

    test.each(tests)(
        'for s = %s, numRows = %d, should return %s',
        (s: string, numRows: number, expected: string) => {
            expect(zigzagConversion(s, numRows)).toEqual(expected);
        }
    );
});

File: problems/problem_004/golang/solution_test.go
package problem_004

import "testing"

type zigzagConversionTest struct {
	s string
	numRows int
	expected string
}

var zigzagConversionTests = []zigzagConversionTest{
	{"HELLOPROGRAMMING", 4, "HRMEPOMILOGANLRG"},
	{"PROBLEMCHALLENGE", 5, "PHRCAEOMLGBELNLE"},
	{"ABCDE", 2, "ACEBD"},
}

func Test_zigzagConversion(t *testing.T) {
	for _, test := range zigzagConversionTests {
		if result := zigzagConversion(test.s, test.numRows); result != test.expected {
			t.Errorf("Output %v not equal to expected %v", result, test.expected)
		}
	}
}